File: orgchat/backend/app/utils/file_upload.py
import os
import uuid
from typing import Tuple

import boto3
from botocore.exceptions import BotoCoreError, ClientError
from fastapi import HTTPException, UploadFile

from app.config import settings

UPLOAD_DIR = "uploads"

ALLOWED_MIME_TYPES = {
    "image/jpeg",
    "image/png",
    "image/gif",
    "image/webp",
    "application/pdf",
    "application/msword",
    "application/vnd.openxmlformats-officedocument.wordprocessingml.document",
    "application/zip",
    "video/mp4",
    "audio/mpeg",
    "audio/mp4",
}

# Magic-byte signatures: (prefix_bytes, expected_mime_or_None_for_multi)
_MAGIC: list[tuple[bytes, str | None]] = [
    (b"\xff\xd8\xff", "image/jpeg"),
    (b"\x89PNG\r\n\x1a\n", "image/png"),
    (b"GIF87a", "image/gif"),
    (b"GIF89a", "image/gif"),
    (b"RIFF", None),              # could be WebP (checked separately)
    (b"%PDF", "application/pdf"),
    (b"PK\x03\x04", None),        # ZIP / DOCX / XLSX
    (b"ID3", "audio/mpeg"),
    (b"\xff\xfb", "audio/mpeg"),
    (b"\xff\xf3", "audio/mpeg"),
    (b"\xff\xf2", "audio/mpeg"),
]

_MULTI_TYPE_ALLOW = {
    # RIFF containers
    "image/webp",
    "audio/mp4",
    # ZIP-based
    "application/zip",
    "application/vnd.openxmlformats-officedocument.wordprocessingml.document",
    "application/msword",
    # MP4 magic starts with 0x00 0x00 0x00 and varies
    "video/mp4",
}


def _detect_mime(header: bytes, declared: str) -> bool:
    """Return True if the file header is consistent with the declared MIME type."""
    for magic, expected in _MAGIC:
        if header.startswith(magic):
            if expected is None:
                # Ambiguous prefix — accept only types we whitelist for these
                return declared in _MULTI_TYPE_ALLOW
            return expected == declared

    # MP4 / MOV: ftyp box at offset 4
    if len(header) >= 8 and header[4:8] in (b"ftyp", b"moov"):
        return declared in ("video/mp4", "audio/mp4")

    # Unknown signature — reject unknown types, accept whitelisted ones
    return declared in _MULTI_TYPE_ALLOW


def _safe_filename(original: str | None) -> str:
    ext = ""
    if original:
        _, ext = os.path.splitext(original)
    return f"{uuid.uuid4().hex}{ext}"


async def validate_and_upload(file: UploadFile) -> Tuple[str, str, int, str]:
    """
    Validate MIME type via magic bytes and size, then upload.
    Returns (file_url, file_name, file_size, mime_type).
    """
    declared_type: str = (file.content_type or "").split(";")[0].strip()

    if declared_type not in ALLOWED_MIME_TYPES:
        raise HTTPException(
            status_code=400,
            detail=f"File type not allowed: {declared_type}",
        )

    content = await file.read()
    file_size = len(content)
    max_bytes = settings.MAX_FILE_SIZE_MB * 1024 * 1024

    if file_size > max_bytes:
        raise HTTPException(
            status_code=400,
            detail=f"File exceeds maximum size of {settings.MAX_FILE_SIZE_MB} MB",
        )

    if file_size == 0:
        raise HTTPException(status_code=400, detail="Empty file")

    header = content[:16]
    if not _detect_mime(header, declared_type):
        raise HTTPException(
            status_code=400,
            detail="File content does not match its declared MIME type",
        )

    safe_name = _safe_filename(file.filename)
    original_name = file.filename or safe_name

    if settings.s3_configured:
        file_url = await _upload_to_s3(content, safe_name, declared_type)
    else:
        file_url = _save_locally(content, safe_name)

    return file_url, original_name, file_size, declared_type


async def _upload_to_s3(content: bytes, key: str, content_type: str) -> str:
    try:
        client = boto3.client(
            "s3",
            region_name=settings.S3_REGION,
            aws_access_key_id=settings.S3_ACCESS_KEY,
            aws_secret_access_key=settings.S3_SECRET_KEY,
        )
        client.put_object(
            Bucket=settings.S3_BUCKET,
            Key=f"uploads/{key}",
            Body=content,
            ContentType=content_type,
        )
        return (
            f"https://{settings.S3_BUCKET}.s3.{settings.S3_REGION}"
            f".amazonaws.com/uploads/{key}"
        )
    except (BotoCoreError, ClientError) as exc:
        raise HTTPException(status_code=500, detail=f"S3 upload failed: {exc}")


def _save_locally(content: bytes, filename: str) -> str:
    os.makedirs(UPLOAD_DIR, exist_ok=True)
    path = os.path.join(UPLOAD_DIR, filename)
    with open(path, "wb") as fh:
        fh.write(content)
    return f"/uploads/{filename}"

File: backend-go/utils/file_upload.go
package utils

import (
	"bytes"
	"context"
	"encoding/base64"
	"errors"
	"fmt"
	"image"
	"image/color"
	"image/jpeg"
	_ "image/gif"
	_ "image/png"
	"io"
	"mime/multipart"
	"net/http"
	"os"
	"path/filepath"
	"strings"

	"github.com/aws/aws-sdk-go-v2/aws"
	awsconfig "github.com/aws/aws-sdk-go-v2/config"
	"github.com/aws/aws-sdk-go-v2/credentials"
	"github.com/aws/aws-sdk-go-v2/service/s3"
	"github.com/google/uuid"
	"github.com/orgchat/backend/config"
)

var allowedMIMETypes = map[string]bool{
	"image/jpeg": true, "image/png": true, "image/gif": true,
	"image/webp": true,
	"application/pdf": true,
	"application/msword": true,
	"application/vnd.openxmlformats-officedocument.wordprocessingml.document": true,
	"application/vnd.ms-excel":                                                  true,
	"application/vnd.openxmlformats-officedocument.spreadsheetml.sheet":         true,
	"application/vnd.ms-powerpoint":                                              true,
	"application/vnd.openxmlformats-officedocument.presentationml.presentation": true,
	"application/zip":         true,
	"application/x-zip-compressed": true,
	"text/plain":              true,
	"text/csv":                true,
	"audio/mpeg":              true,
	"audio/wav":               true,
	"audio/ogg":               true,
	"audio/webm":              true,
	"video/mp4":               true,
	"video/webm":              true,
	"video/ogg":               true,
}

func detectMIME(header []byte) string {
	switch {
	case len(header) >= 3 && header[0] == 0xFF && header[1] == 0xD8 && header[2] == 0xFF:
		return "image/jpeg"
	case len(header) >= 4 && bytes.Equal(header[:4], []byte{0x89, 0x50, 0x4E, 0x47}):
		return "image/png"
	case len(header) >= 3 && bytes.Equal(header[:3], []byte{0x47, 0x49, 0x46}):
		return "image/gif"
	case len(header) >= 12 && bytes.Equal(header[:4], []byte{0x52, 0x49, 0x46, 0x46}) && bytes.Equal(header[8:12], []byte{0x57, 0x45, 0x42, 0x50}):
		return "image/webp"
	case len(header) >= 4 && bytes.Equal(header[:4], []byte{0x25, 0x50, 0x44, 0x46}):
		return "application/pdf"
	case len(header) >= 4 && bytes.Equal(header[:4], []byte{0x50, 0x4B, 0x03, 0x04}):
		return "application/zip"
	case len(header) >= 3 && (bytes.Equal(header[:2], []byte{0xFF, 0xFB}) || bytes.Equal(header[:3], []byte{0x49, 0x44, 0x33})):
		return "audio/mpeg"
	case len(header) >= 8 && bytes.Equal(header[4:8], []byte{0x66, 0x74, 0x79, 0x70}):
		return "video/mp4"
	}
	return ""
}

type UploadResult struct {
	URL       string `json:"url"`
	FileName  string `json:"file_name"`
	FileSize  int64  `json:"file_size"`
	MIMEType  string `json:"mime_type"`
	Thumbnail string `json:"thumbnail"` // base64 data URL for images; empty for non-images
}

// generateThumbnail decodes an image and returns a tiny base64 JPEG data URL
// (≤40px wide) suitable for use as a blurred placeholder when the file expires.
// Returns "" if the MIME type is not a supported image or decoding fails.
func generateThumbnail(data []byte, mimeType string) string {
	switch mimeType {
	case "image/jpeg", "image/png", "image/gif", "image/webp":
	default:
		return ""
	}
	src, _, err := image.Decode(bytes.NewReader(data))
	if err != nil {
		return ""
	}
	b := src.Bounds()
	origW, origH := b.Dx(), b.Dy()
	if origW == 0 || origH == 0 {
		return ""
	}
	const maxW = 40
	thumbW := maxW
	thumbH := (origH * thumbW) / origW
	if thumbH < 1 {
		thumbH = 1
	}
	// Nearest-neighbour downscale — no external deps required.
	dst := image.NewNRGBA(image.Rect(0, 0, thumbW, thumbH))
	for y := range thumbH {
		for x := range thumbW {
			sx := b.Min.X + (x*origW)/thumbW
			sy := b.Min.Y + (y*origH)/thumbH
			r, g, bl, a := src.At(sx, sy).RGBA()
			dst.SetNRGBA(x, y, color.NRGBA{
				R: uint8(r >> 8),
				G: uint8(g >> 8),
				B: uint8(bl >> 8),
				A: uint8(a >> 8),
			})
		}
	}
	var buf bytes.Buffer
	if err := jpeg.Encode(&buf, dst, &jpeg.Options{Quality: 20}); err != nil {
		return ""
	}
	return "data:image/jpeg;base64," + base64.StdEncoding.EncodeToString(buf.Bytes())
}

func ValidateAndUpload(file *multipart.FileHeader, declaredMIME string) (*UploadResult, error) {
	maxBytes := config.App.MaxFileSizeMB * 1024 * 1024
	if file.Size > maxBytes {
		return nil, fmt.Errorf("file size %d exceeds limit of %dMB", file.Size, config.App.MaxFileSizeMB)
	}
	if !allowedMIMETypes[declaredMIME] {
		return nil, errors.New("file type not allowed")
	}

	f, err := file.Open()
	if err != nil {
		return nil, err
	}
	defer f.Close()

	data, err := io.ReadAll(f)
	if err != nil {
		return nil, err
	}

	if len(data) >= 4 {
		// First try magic-byte detection for binary formats.
		detected := detectMIME(data[:min(len(data), 12)])
		if detected == "" {
			// Fall back to the stdlib sniffer (reads up to 512 bytes).
			detected = http.DetectContentType(data[:min(len(data), 512)])
			// http.DetectContentType always returns something; normalise the
			// generic fallback so it doesn't trigger a false mismatch.
			if detected == "application/octet-stream" {
				detected = ""
			}
		}
		if detected != "" && detected != declaredMIME {
			return nil, errors.New("file content does not match declared type")
		}
	}

	ext := filepath.Ext(file.Filename)
	safeFilename := uuid.New().String() + ext
	thumbnail := generateThumbnail(data, declaredMIME)

	if config.App.S3Bucket != "" {
		url, err := uploadToS3(data, safeFilename, declaredMIME)
		if err != nil {
			return nil, err
		}
		return &UploadResult{URL: url, FileName: file.Filename, FileSize: file.Size, MIMEType: declaredMIME, Thumbnail: thumbnail}, nil
	}
	return saveLocally(data, safeFilename, file.Filename, file.Size, declaredMIME, thumbnail)
}

func uploadToS3(data []byte, key, contentType string) (string, error) {
	cfg, err := awsconfig.LoadDefaultConfig(context.Background(),
		awsconfig.WithRegion(config.App.S3Region),
		awsconfig.WithCredentialsProvider(credentials.NewStaticCredentialsProvider(
			config.App.AWSAccessKeyID,
			config.App.AWSSecretAccessKey,
			"",
		)),
	)
	if err != nil {
		return "", err
	}
	client := s3.NewFromConfig(cfg)
	_, err = client.PutObject(context.Background(), &s3.PutObjectInput{
		Bucket:      aws.String(config.App.S3Bucket),
		Key:         aws.String(key),
		Body:        bytes.NewReader(data),
		ContentType: aws.String(contentType),
	})
	if err != nil {
		return "", err
	}
	return fmt.Sprintf("https://%s.s3.%s.amazonaws.com/%s", config.App.S3Bucket, config.App.S3Region, key), nil
}

func saveLocally(data []byte, safeFilename, originalName string, size int64, mimeType, thumbnail string) (*UploadResult, error) {
	dir := config.App.UploadsDir
	if err := os.MkdirAll(dir, 0755); err != nil {
		return nil, err
	}
	dest := filepath.Join(dir, safeFilename)
	if err := os.WriteFile(dest, data, 0644); err != nil {
		return nil, err
	}
	url := "/uploads/" + safeFilename
	uploadsBase := strings.TrimPrefix(dir, "./")
	_ = uploadsBase
	return &UploadResult{URL: url, FileName: originalName, FileSize: size, MIMEType: mimeType, Thumbnail: thumbnail}, nil
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

// DeleteFile removes a previously uploaded file from local disk or S3.
// fileURL is the value stored in messages.file_url.
func DeleteFile(fileURL string) error {
	if config.App.S3Bucket != "" {
		// S3 URL: https://<bucket>.s3.<region>.amazonaws.com/<key>
		parts := strings.SplitN(fileURL, ".amazonaws.com/", 2)
		if len(parts) != 2 {
			return fmt.Errorf("unexpected S3 URL format: %s", fileURL)
		}
		key := parts[1]
		cfg, err := awsconfig.LoadDefaultConfig(context.Background(),
			awsconfig.WithRegion(config.App.S3Region),
			awsconfig.WithCredentialsProvider(credentials.NewStaticCredentialsProvider(
				config.App.AWSAccessKeyID,
				config.App.AWSSecretAccessKey,
				"",
			)),
		)
		if err != nil {
			return err
		}
		client := s3.NewFromConfig(cfg)
		_, err = client.DeleteObject(context.Background(), &s3.DeleteObjectInput{
			Bucket: aws.String(config.App.S3Bucket),
			Key:    aws.String(key),
		})
		return err
	}
	// Local URL: /uploads/<filename>
	filename := strings.TrimPrefix(fileURL, "/uploads/")
	dest := filepath.Join(config.App.UploadsDir, filename)
	if err := os.Remove(dest); err != nil && !os.IsNotExist(err) {
		return err
	}
	return nil
}
